### PR TITLE
Multiple networks support(part 3): new payload version -39

### PR DESCRIPTION
**Pull request by Jack-Works**
This payload indicates a new way to search for keys on the gun DB. The new way is designed to keep the user's privacy safe.

The design is documented there.

https://github.com/DimensionDev/Maskbook/wiki/Data-structure-on-Gun-version-2

- [ ] Select a suitable hash function,  iterations, and N to cut chars.

------------------------------------------
By merging this pr, network provider must select a range of acceptable payload.

For facebook.com, v40 and v39 is accepable (for compabitablility). For other networks, only v39 is supported. And on Facebook.com there is no way to create old v40 payload anymore.

### CODE REVIEW

File: src/utils/memoize.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    f: T,
+    resolver: (...args: Args) => any = ((a: any) => a) as any,
+) {
+    const memorizedFunction = memoize(
+        (async function(...args: Args) {
+            try {
+                // ? DO NOT remove "await" here
+                return await f(...args)
+            } catch (e) {
+                memorizedFunction.cache.delete(resolver(...args))
```
Comment (SunriseFox) [inline src/utils/memoize.ts]: resolver is possibly undefined here?... Typescript is not reporting this so I'm wondering if I got it wrong...

File: src/network/gun/version.2/post.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                .map()
+                .on((data: PostOnGun2) => {
+                    const data2 = Object.assign({}, data)
+                    delete data2._
+                    callback(data2)
+                })
+        })
+    })
+    // tslint:disable-next-line: no-parameter-reassignment
+    return () => (callback = () => {})
```
Comment (SunriseFox) [inline src/network/gun/version.2/post.ts]: what does this (and this function) do...

File: src/network/gun/version.2/post.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    hashPostSalt(postSalt).then(postHash => {
+        hashCryptoKey(partitionByCryptoKey).then(keyHash => {
+            gun2.get(postHash)
+                // @ts-ignore
+                .get(keyHash)
+                .map()
+                .on((data: PostOnGun2) => {
+                    const data2 = Object.assign({}, data)
+                    delete data2._
+                    callback(data2)
```
Comment (SunriseFox) [inline src/network/gun/version.2/post.ts]: `const {_, ...data2} = data` maybe ok...?

File: src/network/gun/version.2/people.ts [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    hashPersonIdentifier(user).then(hash =>
+        gun2.get(hash).on((data: PersonOnGun2) => {
+            const data2 = Object.assign({}, data)
+            // @ts-ignore
+            delete data2._
+            callback(data2)
+        }),
+    )
+    // tslint:disable-next-line: no-parameter-reassignment
+    return () => (callback = () => {})
```
Comment (SunriseFox) [inline src/network/gun/version.2/people.ts]: What does it do...

File: src/extension/background-script/prepareOthersKeyForEncryption.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+ */
+export async function prepareOthersKeyForEncryptionV40(
+    to: PersonIdentifier[],
+): Promise<
+    {
+        name: string
+        key: CryptoKey
+    }[]
+> {
+    const data = await Promise.all(to.map(x => x).map(queryPersonDB))
```
Comment (SunriseFox) [inline src/extension/background-script/prepareOthersKeyForEncryption.ts]: what does it do by map x => x...? maybe filter?

File: src/extension/background-script/prepareOthersKeyForEncryption.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    return data
+        .filter((x): x is NonNullable<typeof x> => !!x)
+        .map(x => ({ name: x.identifier.userId, key: x.publicKey! }))
+}
+/**
+ * @internal
+ */
+export async function prepareOthersKeyForEncryptionV39(to: PersonIdentifier[]): Promise<CryptoKey[]> {
+    const data = await Promise.all(to.map(x => x).map(queryPersonDB))
```
Comment (SunriseFox) [inline src/extension/background-script/prepareOthersKeyForEncryption.ts]: maybe `to.filter` ?

File: src/extension/background-script/CryptoServices/getSharedListOfPost.ts [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        // tslint:disable-next-line: deprecation
+        const post = await Gun1.gun1
+            .get('posts')
+            .get(postSalt)
+            .once().then!()
+        if (!post) return []
+        delete post._
+        const nameInGun = Object.keys(post)
+        // ? version 40 is for old facebook only
+        nameInGun.forEach(x => new PersonIdentifier('facebook.com', x))
```
Comment (SunriseFox) [inline src/extension/background-script/CryptoServices/getSharedListOfPost.ts]: seems no side effects by calling `new PersonIdentifier`

File: src/utils/type-transform/Payload.ts [hunk 8]
```diff
@@ -1,18 +1,22 @@
...
- * Detect if there is version -40 payload
+ * Detect if there is version -40 or -39 payload
  */
-function deconstructAlpha40(str: string, throws = false): PayloadAlpha40 | null {
-    // 🎼2/4|ownersAESKeyEncrypted|iv|encryptedText|signature:||
-    const [_, payloadStart] = str.split('🎼2/4|')
+function deconstructAlpha40_Or_Alpha39(str: string, throws = false): Payload | null {
+    // ? payload is 🎼2/4|ownersAESKeyEncrypted|iv|encryptedText|signature:||
+    // ? payload is 🎼3/4|ownersAESKeyEncrypted|iv|encryptedText|signature:||
+    const unified = str.replace('🎼2/4', '🎼3/4')
```
[the remark this answers is not in the document]
Comment (Jack-Works) [reply src/utils/type-transform/Payload.ts]: It's a mistake. Fixed.

File: src/utils/memoize.ts (hunk 1 above)
Comment (Jack-Works) [reply src/utils/memoize.ts]: It has a default value `((a: any) => a) as any`

File: src/network/gun/version.2/post.ts (hunk 2 above)
Comment (Jack-Works) [reply src/network/gun/version.2/post.ts]: I want to remove the listener but the gun's `off` has bugs. 
This line replaces the original `callback` and this new `callback` doesn't hold any reference at the current closure so no memory will be leaked.

File: src/network/gun/version.2/people.ts (hunk 4 above)
Comment (Jack-Works) [reply src/network/gun/version.2/people.ts]: Explained above

**Pull request merged**